### PR TITLE
TYP: pandas/core/dtypes/base.py

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/core/dtypes/base.py [hunk 1]
```diff
@@ -192,7 +207,40 @@ def construct_array_type(cls):
...
+        Returns
+        -------
+        dict
+        """
+        if not isinstance(string, str):
+            raise TypeError(
+                f"'construct_from_string' expects a string, got {type(string)}"
+            )
+
+        if hasattr(cls, "_match"):
```
Comment (jreback) [inline pandas/core/dtypes/base.py]: what uses this?

Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: some repeated logic in pandas/core/dtypes/dtypes.py can be removed with this. will remove for now til needed. (will also be adding a test in the base extension tests for consistent error messages for wrong type to construct_from_string.)

I'll open a PR to break off changes to pandas/core/dtypes/dtypes.py without the deduplication for now.

File: pandas/core/dtypes/base.py [hunk 2]
```diff
@@ -119,11 +121,11 @@ def __eq__(self, other: Any) -> bool:
     def __hash__(self) -> int:
         return hash(tuple(getattr(self, attr) for attr in self._metadata))
 
-    def __ne__(self, other) -> bool:
+    def __ne__(self, other: Any) -> bool:
```
Comment (WillAyd) [inline pandas/core/dtypes/base.py]: Did we ever decide on consistent use of Any vs object here? Thought we were using object but maybe missed that

**Review by WillAyd — COMMENTED** [review]
lgtm one question

File: pandas/core/dtypes/base.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/base.py]: in this class we already have `def __eq__(self, other: Any) -> bool:` and in #29947, `Any` was applied throughout.

I think it should be `object` in order to check the body of the ``__eq__``/``__ne__`` methods, but since we want consistency these will need to be changed as well...

\_\_eq__
```
pandas/core/arrays/interval.py:    def __eq__(self, other):
pandas/core/arrays/sparse/dtype.py:    def __eq__(self, other: Any) -> bool:
pandas/core/arrays/_arrow_utils.py:        def __eq__(self, other):
pandas/core/arrays/_arrow_utils.py:        def __eq__(self, other):
pandas/core/dtypes/base.py:    def __eq__(self, other: Any) -> bool:
pandas/core/dtypes/dtypes.py:    def __eq__(self, other: Any) -> bool:
pandas/core/dtypes/dtypes.py:    def __eq__(self, other: Any) -> bool:
pandas/core/dtypes/dtypes.py:    def __eq__(self, other: Any) -> bool:
pandas/core/dtypes/dtypes.py:    def __eq__(self, other: Any) -> bool:
pandas/core/indexes/frozen.py:    def __eq__(self, other: Any) -> bool:
pandas/io/pytables.py:    def __eq__(self, other: Any) -> bool:
pandas/io/pytables.py:    def __eq__(self, other: Any) -> bool:
pandas/io/stata.py:    def __eq__(self, other: Any) -> bool:
pandas/tests/groupby/test_function.py:        def __eq__(self, other):
pandas/tests/indexing/test_indexing.py:            def __eq__(self, other) -> bool:
pandas/tests/scalar/timedelta/test_timedelta.py:            def __eq__(self, other):
pandas/tests/scalar/timestamp/test_comparisons.py:        def __eq__(self, other) -> bool:
pandas/tests/series/test_ufunc.py:        def __eq__(self, other) -> bool:
pandas/tests/test_algos.py:            def __eq__(self, other) -> bool:
pandas/tseries/offsets.py:    def __eq__(self, other: Any) -> bool:
pandas/_libs/tslibs/offsets.pyx:    def __eq__(self, other: Any) -> bool:
```
\_\_ne__
```
pandas/core/arrays/interval.py:    def __ne__(self, other):
pandas/io/pytables.py:    def __ne__(self, other) -> bool:
pandas/tests/series/test_duplicates.py:        def __ne__(self, other):
pandas/tseries/offsets.py:    def __ne__(self, other):
pandas/_libs/tslibs/offsets.pyx:    def __ne__(self, other):
```

However, doing this raises issues with Extension Array equalities, see #31646. So should be done in a separate PR, I think though that this does affirm the need to use `object` instead of `Any` for more thorough checking.

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -74,7 +74,7 @@ def register(self, dtype: Type[ExtensionDtype]) -> None:
 
     def find(
         self, dtype: Union[Type[ExtensionDtype], str]
-    ) -> Optional[Type[ExtensionDtype]]:
+    ) -> Optional[Union[ExtensionDtype, Type[ExtensionDtype]]]:
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py]: why is this changed?

**Review by WillAyd — APPROVED** [review]
lgtm outside of Jeff's question

File: pandas/core/dtypes/dtypes.py (hunk 3 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/dtypes.py]: construct_from_string on L98 returns an instance. However, since the type annotations for construct_from_string were removed following https://github.com/pandas-dev/pandas/pull/31352#discussion_r371590898, this can also be removed for now.

**Comment by jbrockmendel** [comment]
This looks fine, any sticking points?

**Comment by WillAyd** [comment]
Thanks @simonjayhawkins
